src/tracker.rs

```diff
@@ -0,0 +1,300 @@
+/// ).with_country(Country::It).with_tracker_configuration(tracker_config);
+/// Datadog::init(configuration).unwrap();
+/// ```
+pub struct TagTrackerConfiguration {
+    count_threshold: usize,
+    actions: Vec<ThresholdAction>,
+}
+
+impl Default for TagTrackerConfiguration {
+    fn default() -> Self {
+        Self {
+            count_threshold: DEFAULT_TAG_THRESHOLD,
+            actions: Vec::new(),
+        }
+    }
+}
+
+impl TagTrackerConfiguration {
+    pub fn new() -> Self {
+        Self {
+            count_threshold: DEFAULT_TAG_THRESHOLD,
+            actions: Vec::new(),
+        }
+    }
+
+    /// Configure the tracker to emit an event when the threshold is reached.
+    /// The count of unique tag sets, per metric, is provided as the tags
+    /// for the event, i.e. for a given metric `metric`, there will be a tag `metric:count`,
+    /// where count is the number of unique tag sets seen for that metric.
+    /// Any number of events may be configured, and all will be emitted when the
+    /// threshold is reached.
+    pub fn with_event(mut self, title: String, text: String) -> Self {
+        self.actions.push(ThresholdAction::Event { title, text });
+        self
+    }
+
+    /// Add a custom action to execute when the custom metric threshold is reached.
+    /// These actions are run exactly once, at the point the threshold is reached.
+    /// Any number of actions may be added.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// prima_datadog::TagTrackerConfiguration::new().with_custom_action(|metric, tags| {
+    ///     println!("Exceeded custom metric threshold for metric {} with tags {:?}", metric, tags);
+    /// });
+    /// ```
+    pub fn with_custom_action(mut self, custom_action: impl FnMut(&str, &[&str]) + Send + Sync + 'static) -> Self {
+        self.actions
+            .push(ThresholdAction::Custom(Box::new(custom_action) as Box<_>));
+        self
+    }
+
+    /// Set the threshold for the maximum number of custom metrics
+    /// This defaults to ```DEFAULT_TAG_THRESHOLD```
+    ///
+    /// The threshold is the maximum number of "custom metrics" which can
+    /// be generated before the tracker actions are run.
+    ///
+    /// A "custom metric" is defined as the unique combination of metric name
+    /// and unique tag set, e.g., a metric, "test", with the following tag sets:
+    /// - [a, b, c]
+    /// - [a, b, d]
+    /// - [a, c, d]
+    /// is counted as 3 "custom metrics", as there are 3 unique tag sets here.
+    ///
+    /// See https://docs.datadoghq.com/developers/metrics/custom_metrics/ for
+    /// more information.
+    pub fn with_threshold(mut self, count_threshold: usize) -> Self {
+        self.count_threshold = count_threshold;
+        self
+    }
+
+    pub(crate) fn build(self) -> Tracker {
+        Tracker::new(self.count_threshold, self.actions)
+    }
+}
+
+/// An iterator over tags, which can be rewound to the beginning.
+struct RewindableTagsIter<'a> {
```

Comment:
I'm a bit uncertain on why we need this iterator... I see those `rewind` calls scattered all over the place but we are only iterating it when we call `.any`/`.map` on it, right? Can't we just have a slice/Vec and call `.iter().map()` on it?

Reply:
Would be nice... unfortunately we can't pass in a slice of `&[S]` where `S: AsRef<str>` because that risks not monomorphizing, as `S` would have to be generic. We also can't clone a `dyn Iterator` because it's `!Sized` :(

This was the best thing I could come up with. I could probably make it much less confusing by adding an `.iter()` method instead of a `.rewind()` method, which would look far more idiomatic.

Reply:
I agree with the criticism here - I don't like the rewind method so much because its possible to "hold wrong" (we have no way to guarantee at the type system level that it's called correctly), making this iterator hand out refs that automatically reset it would be nice, or making it `copy` with an `Rc` underlying iterator, or similar

Reply:
See above, removed for now